### ISSUE TITLE
Investigate gateway timeout

### ISSUE THREAD

**Issue by Ingvord** [tango-controls/rest-server#129]
Originally reported by: **Ingvord (Bitbucket: [Ingvord](https://bitbucket.org/Ingvord), GitHub: [Ingvord](https://github.com/Ingvord))**

----------------------------------------

Dear Igor,

Since few days (about one week) the TangoRestServer is hanging up some time to time.
The tango server is answering properly but when trying with Rest, we get a Gateway timeout (see screen dump below).
Each time I solved the problem by simply restarting the TangoRestServer.
We suspect that these crashed can be correlated with the fact that several clients are intensively accessing it recently (you, me and Jérome...)
Do you have an idea about the possible cause?
Can you reproduce it on another instance?  if not I can give you remote access to my instance to possibly track the problem

Cheers

Jean-Michel

> -------- Forwarded Message --------
> Subject:	Re: Facade Rest API ne répond pas ce matin
> Date:	Wed, 12 Apr 2017 10:10:16 +0200
> From:	Jerome Cambon
> To:	Jean-Michel CHAIZE
> 
> Bonjour Jean-Michel,
> 
> Même problème ce matin.
> Je me demande si ces problèmes ne sont pas accentués par le fait que depuis quelques jours on fait tous (donc 4) des accès assez intensifs à l’interface REST Facade ?
> 
> Bonne journée,
> Jérome
> PS: peux-t-on t’appeler aujourd’hui, pour régler notamment des problèmes de logistique pour Florence
> 
> 
> > On Apr 10, 2017, at 11:03 AM, Jerome Cambon wrote:
> > 
> > Bonjour Jean-Michel,
> > 
> > Je ne sais pas si tu es au courant, mais l’interface REST de Facade ne répond pas :
> > 
> > $ curl -u "https://mstatus.esrf.fr/tango/rest/rc4/hosts/tangorest01.esrf.fr/10000/devices/sys/mcs/facade/attributes/current/value" | python -m json.tool
> > 
> >   % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
> >                                  Dload  Upload   Total   Spent    Left  Speed
> > 
> > 
> > 100    42    0    42    0     0      0      0 --:--:--  0:00:50 --:--:--     9
> > 
> > {
> >     "message": "504: Gateway Time-out"
> > }
> > 
> > 
> > 
> > Merci,
> > Jerome Cambon

----------------------------------------
- Bitbucket: https://bitbucket.org/hzgwpn/mtangorest.server/issue/129

**Comment by Ingvord** [comment]
*Original comment by* **Ingvord (Bitbucket: [Ingvord](https://bitbucket.org/Ingvord), GitHub: [Ingvord](https://github.com/Ingvord))**:

----------------------------------------

OK,

that was rather easy.

By default Tomcat allows only 200 connections to be opened simultaneously. When this limit is reached no more incoming connection are allowed i.e. timeout.

Here I have created a simple project to reproduce this behavior: https://github.com/Ingvord/test-tomcat-blocked

But why is this happening? The problem is with events, specifically I block incoming thread till TANGO sends a new value. That why all http-nio-xxx threads are in BLOCKED state due to EventHelper.get method:

```
$ sudo /sware/pub/java-1.8.0_91/linux_x64/bin/jstack -F 18116 | grep "EventHelper.get(long)" | wc -l
200
```

So there is a number of questions:

1. Why TANGO does not send updates?
2. To fix this only on the server side I can only decrease time out of EventHelper.get and return null if there is no update (This I will do now as easiestr and quickest possible solution)
3. Support from both sides is required if we want efficient event system implementation i.e. multiplexing/demultiplexing of events flow so that only one connection is used to handle all events
4. Websockets? 
5. Server side push in HTTP/2?

Any ideas?

Regards,

  Igor.

**Comment by Ingvord** [comment]
*Original comment by* **Ingvord (Bitbucket: [Ingvord](https://bitbucket.org/Ingvord), GitHub: [Ingvord](https://github.com/Ingvord))**:

----------------------------------------

Kind of fixed in 1.2 (default timeout decreased to 3s)
